cmd/nonce-service/main.go

```diff
@@ -15,14 +17,73 @@ type Config struct {
 	NonceService struct {
 		cmd.ServiceConfig
 
-		MaxUsed     int
+		MaxUsed int
+		// TODO(#6610): Remove once we've moved to derivable prefixes by
+		// default.
 		NoncePrefix string
 
+		// UseDerivablePrefix indicates whether or not to use a nonce prefix
+		// derived from the SHA256 of the host's IP address and a salt.
+		//
+		// TODO(#6610): Remove once we've moved to derivable prefixes by
+		// default.
+		UseDerivablePrefix bool
+		// PrefixSalt is the path to a file containing the salt used to derive
+		// the nonce prefix. This is only used if UseDerivablePrefix is true.
+		//
+		// TODO(#6610): Edit this comment once we've moved to derivable prefixes
+		// by default.
+		PrefixSalt cmd.PasswordConfig
+
 		Syslog  cmd.SyslogConfig
 		Beeline cmd.BeelineConfig
 	}
 }
 
+func derivePrefix(salt string, grpcAddr string) (string, error) {
+	addrs, err := net.InterfaceAddrs()
+	if err != nil {
+		return "", err
+	}
+
+	// If the configuration specifies an IPv4 address and port, we should use it
+	// to derive the nonce prefix.
+	host, port, err := net.SplitHostPort(grpcAddr)
+	if err != nil {
+		return "", fmt.Errorf("parsing gRPC listen address: %w", err)
+	}
+	if host != "" && port != "" {
+		hostIP := net.ParseIP(host)
+		if hostIP != nil && hostIP.To4() != nil {
+			return nonce.DerivePrefix(grpcAddr, salt), nil
+		}
+	}
+
+	// Otherwise, we should use the only non-loopback interface with an IPv4
+	// address we find on the system and the port passed in the gRPC
+	// configuration.
```

Comment:
There are a few situations which this approach is attempting to address. All of them are working around how `net.Listen()` will use the address from `cmd.ServiceConfig.GRPC.Address`, passed on line `163` inside of `grpc.Build()`.

#### Address is set to `:1234`
The RPC server will bind port `1234` on all interfaces including loopback. This is equivalent to `0.0.0.0:1234` for IPv4 or `[::]:1234` if the server is configured to support IPv6.
  - In production this is fine. We expect our deployment targets to have just one non-loopback interface. It's safe for us to assume that this is the IP:port we're going to derive the prefix from on the client side.
  - In boulder-dev this isn't fine. We have several non-loopback interfaces and we want to bind a `nonce-server` on port `9101` of each `10.77.77.77` and `10.88.88.88`. `net.Listen()` is going to attempt to bind `9101` on all interfaces, this will succeed for `nonce1` and `nonce2` will fail to start.

To solve for the latter situation I've required that the operator specify an interface IP when multiple interfaces are present.

#### Address is set to `nonce1.service.consul:1234`
The RPC server, via `net.Listen()`, will attempt to lookup the DNS name `nonce1.service.consul`. If this fails, the nonce-service will panic with the message `...listen tcp: lookup nonce1.service.consul: no such host`. Consul will respond with `10.77.77.77` and the RPC server will bind port `1234` on the `10.77.77.77` interface. Unfortunately all of this takes place during the construction of the RPC client. We must know the prefix when we construct the `nonce.NonceService` object we pass to the RPC client constructor. To my knowledge this gives us three options:
  - (current) Force the operator to specify one of the interfaces by the IP address.
  - Duplicate the DNS lookup logic from `net.Listen()` and ensure that either an IP or hostname is passed in the address field.
  - Possibly, allow the prefix to be passed after the RPC server is constructed but before we call `cmd.FailOnError(start())`. This feels pretty messy and would mean we're modifying a nicely generalized `grpc.Build()` for the sake of `nonce-service`.

Reply:
> I agree. It's common to have multiple interfaces on a host with different IP addresses. The property this code requires is that the IP/port we're hashing is that same IP/port that is discoverable by the WFE when doing SRV lookups. We should make that explicit in the documentation, and we should change our handling of listen ports so that it _requires_ a specific IP address. I think that's the only way to make this operation reliable, and it would also allow us to delete a lot of this code that tries to find the likely IP address that the service is listening on.

😓 This comment hadn't backfilled before I finished drafting my response to Aaron's comment. Please let me know what you think of my comment above though. I believe we could possibly leave this implementation as it will have the same effect; it will only apply in the case of the nonce service. But it won't necessitate a configuration change on the part of SRE unless absolutely necessary.

Reply:
I checked with SRE and it sounds like all prod services bind a specific IP address (for all services). So if we make the changes in the dev environment to bind a specific IP address, we can start requiring it when we bind a port for a service.

There's one other option: rather than guessing which IP address a request is going to come in on, we can at request time ask which IP address a request _actually_ came in on. Specifically I'm thinking of LocalAddrContextKey from https://pkg.go.dev/net/http#pkg-variables:

	// LocalAddrContextKey is a context key. It can be used in
	// HTTP handlers with Context.Value to access the local
	// address the connection arrived on.
	// The associated value will be of type net.Addr.
	LocalAddrContextKey = &contextKey{"local-addr"}

Oh except gRPC doesn't actually use net/http, it used a forked internal copy for its own purposes. And that copy doesn't seem to have LocalAddrContextKey. 😾

Reply:
Okay, sounds like we're going to rip the guessing out of this and mandate that the operator supply an IP and port to setup an RPC server. I'll do that.